### PR TITLE
[DWB] Option to limit velocity commands in trajectory generator

### PR DESCRIPTION
**Pull request by huiyulhy**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | Issue #3277 |
| Primary OS tested on | Ubuntu 22.04|
| Robotic platform tested on | Turtlebot3 Gazebo simulation - waffle |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points
- Added an option in the standard trajectory generator in DWB to allow the user to limit the velocity command in the trajectory 
-based on current trajectory through a new parameter (`limit_vel_cmd_in_traj_`)
- This option ensures that the acceleration constraints are respected in the event there is a large difference between the current and commanded velocity - which may be more useful when there is no velocity smoother in use
- `limit_vel_cmd_in_traj_` is set to false by default, but when set to true, would use the first velocity computed in the trajectory rollout as the trajectory velocity

Default behavior: 
dev_default

When `limit_vel_cmd_in_traj_` is set to `true`:
proposed_default

We observed slightly improved velocity tracking, and similar time to reach the same goal setting this parameter to true.

## Description of documentation updates required from your changes
- Added parameter `limit_vel_cmd_in_traj_` to DWB local planner (in standard trajectory generator)
---

## Future work that may be required in bullet points
- This parameter could also be extensible to other trajectory generation methods within DWB

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp [hunk 1]
```diff
@@ -75,6 +75,10 @@ void StandardTrajectoryGenerator::initialize(
     nh,
     plugin_name + ".include_last_point", rclcpp::ParameterValue(true));
 
+  nav2_util::declare_parameter_if_not_declared(
+    nh,
+    plugin_name + ".limit_vel_cmd_in_traj_", rclcpp::ParameterValue(false));
```
Comment (SteveMacenski) [inline nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp:80]:
```suggestion
    plugin_name + ".limit_vel_cmd_in_traj", rclcpp::ParameterValue(false));
```

File: nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp [hunk 2]
```diff
@@ -89,6 +93,7 @@ void StandardTrajectoryGenerator::initialize(
   nh->get_parameter(plugin_name + ".linear_granularity", linear_granularity_);
   nh->get_parameter(plugin_name + ".angular_granularity", angular_granularity_);
   nh->get_parameter(plugin_name + ".include_last_point", include_last_point_);
+  nh->get_parameter(plugin_name + ".limit_vel_cmd_in_traj_", limit_vel_cmd_in_traj_);
```
Comment (SteveMacenski) [inline nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp:96]:
```suggestion
  nh->get_parameter(plugin_name + ".limit_vel_cmd_in_traj", limit_vel_cmd_in_traj_);
```

File: nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp [hunk 3]
```diff
@@ -156,9 +161,14 @@ dwb_msgs::msg::Trajectory2D StandardTrajectoryGenerator::generateTrajectory(
   double running_time = 0.0;
   std::vector<double> steps = getTimeSteps(cmd_vel);
   traj.poses.push_back(start_pose);
+  bool firstVel = false;
```
Comment (SteveMacenski) [inline nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp:164]: Variables should be `snake_case` not `camelCase`

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Great! We also need:
- Nav2 migration guide updated to include this change and the new parameter
- Nav2 configuration guide for DWB including this new parameter
- You have a couple of linting errors: https://app.circleci.com/pipelines/github/ros-navigation/navigation2/12645/workflows/2c298f6b-d296-44b0-a985-b33143906d5d/jobs/38206/tests

**Comment by huiyulhy** [comment]
> Great! We also need:
> 
> * Nav2 migration guide updated to include this change and the new parameter
> * Nav2 configuration guide for DWB including this new parameter
> * You have a couple of linting errors: https://app.circleci.com/pipelines/github/ros-navigation/navigation2/12645/workflows/2c298f6b-d296-44b0-a985-b33143906d5d/jobs/38206/tests

Thanks for the suggestion! Am updating for the documentation for Nav2 and I wanted to confirm that the addition of this change in the migration guide would be from Jazzy to K-Turtle?

**Comment by SteveMacenski** [comment]
That is correct!

File: nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp (hunk 1 above)
Comment (huiyulhy) [reply nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp]: Updated this, thanks for catching!

File: nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp (hunk 3 above)
Comment (huiyulhy) [reply nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp]: Updated this, thanks for pointing it out!

File: nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp (hunk 2 above)
Comment (huiyulhy) [reply nav2_dwb_controller/dwb_plugins/src/standard_traj_generator.cpp]: Updated too

**Comment by huiyulhy** [comment]
> Great! We also need:
> 
> * Nav2 migration guide updated to include this change and the new parameter
> * Nav2 configuration guide for DWB including this new parameter
> * You have a couple of linting errors: https://app.circleci.com/pipelines/github/ros-navigation/navigation2/12645/workflows/2c298f6b-d296-44b0-a985-b33143906d5d/jobs/38206/tests

@SteveMacenski thanks for the feedback! Fixed the linting errors and updated the documentation in [this PR](https://github.com/ros-navigation/docs.nav2.org/pull/591) on nav2 docs!

**Pull request merged**